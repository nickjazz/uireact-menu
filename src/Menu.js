import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {isObject, values, get} from 'lodash';
import MenuItem from './MenuItem'
import MenuSub from './MenuSub'
import MenuLabel from './MenuLabel'
import MenuWrapper from './MenuWrapper'
// import './menu.scss'
// import './dashborad.scss'
// import './lightmenu.scss'

let i = 0

class Menu extends Component {

  static propTypes = {
    children: PropTypes.any,
    containerClass: PropTypes.string,
    data: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.array
    ]),
    direction: PropTypes.oneOf([
      'horizontal',
      'vertical'
    ])
  }

  static Item = MenuItem
  static Sub = MenuSub
  static Label = MenuLabel
  static Wrapper = MenuWrapper

  static defaultProps = {
    direction: 'vertical',
    containerClass: 'lightmenu'
  }

  jsonData = null

  componentWillMount() {
    const { data } = this.props;
    if (data && isObject(data)) {
      this.jsonData = values(data)
    }
  }

  renderMenu = (menu) => {
    const submenu = get(menu, ['submenu'])
    return(
      <MenuItem
        // ignore link if have submenu
        to={ submenu ? null : (menu.href || menu.to)}
        key={i++}
      >
        <MenuLabel>{get(menu, 'title')}</MenuLabel>
        { submenu &&
          <MenuSub>
            {
              submenu.map(sub => {
                return this.renderMenu(sub)
              })
            }

          </MenuSub>
        }
      </MenuItem>
    )
  }


  render() {

    const { children, data, direction, containerClass} = this.props
    let jsonDataChild

    if (this.jsonData)
      jsonDataChild = this.jsonData.map((e, i) => {
        return this.renderMenu(e)
      })

    const enhanceChild = data
    ? jsonDataChild
    : children

    return (
      <nav className={`uireact-menu ${direction} ${containerClass}`}>
        {enhanceChild}
      </nav>
    );
  }

}


export default Menu;
