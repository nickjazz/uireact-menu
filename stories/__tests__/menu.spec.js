import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Menu from '../../src';
const menuData = require('../../stories/menu.json')
const menuArray = [{
  'title': 'Menu 1',
  'href': 'google.com',
  'submenu': [{
      'title': 'Sub Menu1',
      'submenu': [{
        'title': 'Sub sub Menu1',
        'href': '#'
      }, {
        'title': 'Sub sub Menu2',
        'href': '#'
      }]
    }, {
      'title': 'Sub Menu2',
      'href': '#'
    }, {
      'title': 'Sub Menu3',
      'href': '#'
    }]
  }, {
    'title': 'Menu 2',
    'submenu': [{
      'title': 'Sub Menu 2',
      'href': '#'
    }]
  }, {
    'title': 'Menu 3',
    'href': '#'
}]

describe('Menu', () => {
  it('Basic usage', () => {
    const output = shallow(
      <div style={{width: '30vw', height: '100vh'}}>
        <Menu>
          <Menu.Item>Home</Menu.Item>
          <Menu.Item>Photos</Menu.Item>
          <Menu.Item>Message</Menu.Item>
          <Menu.Item>Music</Menu.Item>
          <Menu.Item>Video </Menu.Item>
        </Menu>
      </div>
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('Direction usage', () => {
    const output = shallow(
      <div style={{width: '100vw'}}>
        <Menu direction='horizontal'>
          <Menu.Item>Home</Menu.Item>
          <Menu.Item>Photos</Menu.Item>
          <Menu.Item>Message</Menu.Item>
          <Menu.Item>Music</Menu.Item>
          <Menu.Item>Video </Menu.Item>
        </Menu>
      </div>
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('Nested menu item', () => {
    const output = shallow(
      <div style={{width: '80vw'}}>
        <div style={{width: '30vw'}}>
          <Menu>
            <Menu.Item>Home</Menu.Item>
            <Menu.Item>
              <Menu.Label>Photos</Menu.Label>
              <Menu.Sub >
                <Menu.Item>item 2-1</Menu.Item>
                <Menu.Item>item 2-2</Menu.Item>
                <Menu.Item>
                  <Menu.Label>item 2-3</Menu.Label>
                  <Menu.Sub>
                    <Menu.Item>item 2-3-1</Menu.Item>
                    <Menu.Item>
                      <Menu.Label>item 2-3-1</Menu.Label>
                      <Menu.Sub>
                        <Menu.Item>item 2-3-1-1</Menu.Item>
                        <Menu.Item>item 2-3-1-2</Menu.Item>
                        <Menu.Item>item 2-3-1-3</Menu.Item>
                      </Menu.Sub>
                    </Menu.Item>
                    <Menu.Item>item 2-3-2</Menu.Item>
                  </Menu.Sub>
                </Menu.Item>
              </Menu.Sub>
            </Menu.Item>
            <Menu.Item>
              <Menu.Label>Message</Menu.Label>
              <Menu.Sub >
                <Menu.Item>item 2-1</Menu.Item>
                <Menu.Item>item 2-2</Menu.Item>
                <Menu.Item>
                  <Menu.Label>item 2-3</Menu.Label>
                  <Menu.Sub>
                    <Menu.Item>item 2-3-1</Menu.Item>
                    <Menu.Item>
                      <Menu.Label>item 2-3-1</Menu.Label>
                      <Menu.Sub>
                        <Menu.Item>item 2-3-1-1</Menu.Item>
                        <Menu.Item>item 2-3-1-2</Menu.Item>
                        <Menu.Item>item 2-3-1-3</Menu.Item>
                      </Menu.Sub>
                    </Menu.Item>
                    <Menu.Item>item 2-3-2</Menu.Item>
                  </Menu.Sub>
                </Menu.Item>
              </Menu.Sub>
            </Menu.Item>
            <Menu.Item>Music</Menu.Item>
            <Menu.Item>Contact</Menu.Item>
            <Menu.Item>Favorite</Menu.Item>
            <Menu.Item>Follow</Menu.Item>
          </Menu>
        </div>

        <br />

        <Menu direction='horizontal'>
          <Menu.Item to='http://google.com'>item 1</Menu.Item>
          <Menu.Item>
            <Menu.Label>Photos</Menu.Label>
            <Menu.Sub>
              <Menu.Item>item 2-1</Menu.Item>
              <Menu.Item>item 2-2</Menu.Item>
              <Menu.Item>
                <Menu.Label>Message</Menu.Label>
                <Menu.Sub>
                  <Menu.Item>item 2-3-1</Menu.Item>
                  <Menu.Item>
                    <Menu.Label>Video</Menu.Label>
                    <Menu.Sub>
                      <Menu.Item>item 2-3-1-1</Menu.Item>
                      <Menu.Item>item 2-3-1-2</Menu.Item>
                      <Menu.Item>item 2-3-1-3</Menu.Item>
                    </Menu.Sub>
                  </Menu.Item>
                  <Menu.Item>item 2-3-2</Menu.Item>
                </Menu.Sub>
              </Menu.Item>
            </Menu.Sub>
          </Menu.Item>
          <Menu.Item>item 3</Menu.Item>
          <Menu.Item>item 4</Menu.Item>
          <Menu.Item>item 5</Menu.Item>
          <Menu.Item>item 6</Menu.Item>
        </Menu>
      </div>
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('With json data', () => {
    const output = mount((
      <div style={{width: '30vw'}}>
        <Menu data={menuData} />
      </div>
    ));
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('With json data', () => {
    const output = mount((
      <div style={{width: '30vw'}}>
        <Menu data={menuArray} />
      </div>
    ));
    expect(shallowToJson(output)).toMatchSnapshot();
  });

});
