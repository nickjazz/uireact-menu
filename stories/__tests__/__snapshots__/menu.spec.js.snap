// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Basic usage 1`] = `
<div
  style={
    Object {
      "height": "100vh",
      "width": "30vw",
    }
  }
>
  <Menu
    direction="vertical"
  >
    <MenuItem>
      Home
    </MenuItem>
    <MenuItem>
      Photos
    </MenuItem>
    <MenuItem>
      Message
    </MenuItem>
    <MenuItem>
      Music
    </MenuItem>
    <MenuItem>
      Video 
    </MenuItem>
  </Menu>
</div>
`;

exports[`Menu Direction usage 1`] = `
<div
  style={
    Object {
      "width": "100vw",
    }
  }
>
  <Menu
    direction="horizontal"
  >
    <MenuItem>
      Home
    </MenuItem>
    <MenuItem>
      Photos
    </MenuItem>
    <MenuItem>
      Message
    </MenuItem>
    <MenuItem>
      Music
    </MenuItem>
    <MenuItem>
      Video 
    </MenuItem>
  </Menu>
</div>
`;

exports[`Menu Nested menu item 1`] = `
<div
  style={
    Object {
      "width": "80vw",
    }
  }
>
  <div
    style={
      Object {
        "width": "30vw",
      }
    }
  >
    <Menu
      direction="vertical"
    >
      <MenuItem>
        Home
      </MenuItem>
      <MenuItem>
        <MenuLabel>
          Photos
        </MenuLabel>
        <MenuSub>
          <MenuItem>
            item 2-1
          </MenuItem>
          <MenuItem>
            item 2-2
          </MenuItem>
          <MenuItem>
            <MenuLabel>
              item 2-3
            </MenuLabel>
            <MenuSub>
              <MenuItem>
                item 2-3-1
              </MenuItem>
              <MenuItem>
                <MenuLabel>
                  item 2-3-1
                </MenuLabel>
                <MenuSub>
                  <MenuItem>
                    item 2-3-1-1
                  </MenuItem>
                  <MenuItem>
                    item 2-3-1-2
                  </MenuItem>
                  <MenuItem>
                    item 2-3-1-3
                  </MenuItem>
                </MenuSub>
              </MenuItem>
              <MenuItem>
                item 2-3-2
              </MenuItem>
            </MenuSub>
          </MenuItem>
        </MenuSub>
      </MenuItem>
      <MenuItem>
        <MenuLabel>
          Message
        </MenuLabel>
        <MenuSub>
          <MenuItem>
            item 2-1
          </MenuItem>
          <MenuItem>
            item 2-2
          </MenuItem>
          <MenuItem>
            <MenuLabel>
              item 2-3
            </MenuLabel>
            <MenuSub>
              <MenuItem>
                item 2-3-1
              </MenuItem>
              <MenuItem>
                <MenuLabel>
                  item 2-3-1
                </MenuLabel>
                <MenuSub>
                  <MenuItem>
                    item 2-3-1-1
                  </MenuItem>
                  <MenuItem>
                    item 2-3-1-2
                  </MenuItem>
                  <MenuItem>
                    item 2-3-1-3
                  </MenuItem>
                </MenuSub>
              </MenuItem>
              <MenuItem>
                item 2-3-2
              </MenuItem>
            </MenuSub>
          </MenuItem>
        </MenuSub>
      </MenuItem>
      <MenuItem>
        Music
      </MenuItem>
      <MenuItem>
        Contact
      </MenuItem>
      <MenuItem>
        Favorite
      </MenuItem>
      <MenuItem>
        Follow
      </MenuItem>
    </Menu>
  </div>
  <br />
  <Menu
    direction="horizontal"
  >
    <MenuItem
      to="http://google.com"
    >
      item 1
    </MenuItem>
    <MenuItem>
      <MenuLabel>
        Photos
      </MenuLabel>
      <MenuSub>
        <MenuItem>
          item 2-1
        </MenuItem>
        <MenuItem>
          item 2-2
        </MenuItem>
        <MenuItem>
          <MenuLabel>
            Message
          </MenuLabel>
          <MenuSub>
            <MenuItem>
              item 2-3-1
            </MenuItem>
            <MenuItem>
              <MenuLabel>
                Video
              </MenuLabel>
              <MenuSub>
                <MenuItem>
                  item 2-3-1-1
                </MenuItem>
                <MenuItem>
                  item 2-3-1-2
                </MenuItem>
                <MenuItem>
                  item 2-3-1-3
                </MenuItem>
              </MenuSub>
            </MenuItem>
            <MenuItem>
              item 2-3-2
            </MenuItem>
          </MenuSub>
        </MenuItem>
      </MenuSub>
    </MenuItem>
    <MenuItem>
      item 3
    </MenuItem>
    <MenuItem>
      item 4
    </MenuItem>
    <MenuItem>
      item 5
    </MenuItem>
    <MenuItem>
      item 6
    </MenuItem>
  </Menu>
</div>
`;

exports[`Menu With json data 1`] = `
<div
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <Menu
    data={
      Object {
        "menu-0": Object {
          "id": "menu-0",
          "title": "Switzerland",
        },
        "menu-1": Object {
          "id": "menu-1",
          "title": "Canada",
        },
        "menu-10": Object {
          "id": "menu-10",
          "title": "Norway",
        },
        "menu-3": Object {
          "id": "menu-3",
          "title": "United Kingdom",
        },
        "menu-4": Object {
          "id": "menu-4",
          "title": "Japan",
        },
        "menu-5": Object {
          "id": "menu-5",
          "title": "Germany",
        },
        "menu-6": Object {
          "id": "menu-6",
          "title": "Sweden",
        },
        "menu-7": Object {
          "id": "menu-7",
          "title": "United States",
        },
        "menu-8": Object {
          "id": "menu-8",
          "title": "Australia",
        },
        "menu-9": Object {
          "id": "menu-9",
          "title": "France",
        },
      }
    }
    direction="vertical"
  >
    <nav
      className="uireact-menu vertical"
    >
      <MenuItem
        key="0"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Switzerland
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="1"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Canada
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="2"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              United Kingdom
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="3"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Germany
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="4"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Japan
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="5"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Sweden
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="6"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              United States
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="7"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Australia
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="8"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              France
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
      <MenuItem
        key="9"
      >
        <div
          className="uireact-menu-item "
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Norway
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </div>
      </MenuItem>
    </nav>
  </Menu>
</div>
`;

exports[`Menu With json data 2`] = `
<div
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <Menu
    data={
      Array [
        Object {
          "href": "google.com",
          "submenu": Array [
            Object {
              "submenu": Array [
                Object {
                  "href": "#",
                  "title": "Sub sub Menu1",
                },
                Object {
                  "href": "#",
                  "title": "Sub sub Menu2",
                },
              ],
              "title": "Sub Menu1",
            },
            Object {
              "href": "#",
              "title": "Sub Menu2",
            },
            Object {
              "href": "#",
              "title": "Sub Menu3",
            },
          ],
          "title": "Menu 1",
        },
        Object {
          "submenu": Array [
            Object {
              "href": "#",
              "title": "Sub Menu 2",
            },
          ],
          "title": "Menu 2",
        },
        Object {
          "href": "#",
          "title": "Menu 3",
        },
      ]
    }
    direction="vertical"
  >
    <nav
      className="uireact-menu vertical"
    >
      <MenuItem
        key="10"
        to={null}
      >
        <div
          className="uireact-menu-item "
          href={null}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Menu 1
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
          <MenuSub>
            <div
              className="uireact-menu-sub"
            >
              <MenuItem
                key="11"
                to={null}
              >
                <div
                  className="uireact-menu-item "
                  href={null}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <MenuLabel>
                    <span
                      className="uireact-menu-label"
                    >
                      Sub Menu1
                      <span
                        className="uireact-menu-icon"
                      />
                    </span>
                  </MenuLabel>
                  <MenuSub>
                    <div
                      className="uireact-menu-sub"
                    >
                      <MenuItem
                        key="12"
                        to="#"
                      >
                        <a
                          className="uireact-menu-item "
                          href="#"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                        >
                          <MenuLabel>
                            <span
                              className="uireact-menu-label"
                            >
                              Sub sub Menu1
                              <span
                                className="uireact-menu-icon"
                              />
                            </span>
                          </MenuLabel>
                        </a>
                      </MenuItem>
                      <MenuItem
                        key="13"
                        to="#"
                      >
                        <a
                          className="uireact-menu-item "
                          href="#"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                        >
                          <MenuLabel>
                            <span
                              className="uireact-menu-label"
                            >
                              Sub sub Menu2
                              <span
                                className="uireact-menu-icon"
                              />
                            </span>
                          </MenuLabel>
                        </a>
                      </MenuItem>
                    </div>
                  </MenuSub>
                </div>
              </MenuItem>
              <MenuItem
                key="14"
                to="#"
              >
                <a
                  className="uireact-menu-item "
                  href="#"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <MenuLabel>
                    <span
                      className="uireact-menu-label"
                    >
                      Sub Menu2
                      <span
                        className="uireact-menu-icon"
                      />
                    </span>
                  </MenuLabel>
                </a>
              </MenuItem>
              <MenuItem
                key="15"
                to="#"
              >
                <a
                  className="uireact-menu-item "
                  href="#"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <MenuLabel>
                    <span
                      className="uireact-menu-label"
                    >
                      Sub Menu3
                      <span
                        className="uireact-menu-icon"
                      />
                    </span>
                  </MenuLabel>
                </a>
              </MenuItem>
            </div>
          </MenuSub>
        </div>
      </MenuItem>
      <MenuItem
        key="16"
        to={null}
      >
        <div
          className="uireact-menu-item "
          href={null}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Menu 2
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
          <MenuSub>
            <div
              className="uireact-menu-sub"
            >
              <MenuItem
                key="17"
                to="#"
              >
                <a
                  className="uireact-menu-item "
                  href="#"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <MenuLabel>
                    <span
                      className="uireact-menu-label"
                    >
                      Sub Menu 2
                      <span
                        className="uireact-menu-icon"
                      />
                    </span>
                  </MenuLabel>
                </a>
              </MenuItem>
            </div>
          </MenuSub>
        </div>
      </MenuItem>
      <MenuItem
        key="18"
        to="#"
      >
        <a
          className="uireact-menu-item "
          href="#"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <MenuLabel>
            <span
              className="uireact-menu-label"
            >
              Menu 3
              <span
                className="uireact-menu-icon"
              />
            </span>
          </MenuLabel>
        </a>
      </MenuItem>
    </nav>
  </Menu>
</div>
`;
